Fix for issue #28

MVesuviusC opened the pull request:
This changes the behavior of the "enumerate_primers" function to create a dataframe of primers stored as characters and not as factors.

MVesuviusC committed:
Update search.R

Changing the way the data frame is created in the enumerate_primers function to keep primer sequences as characters and not factors

MVesuviusC committed:
Merge pull request #1 from MVesuviusC/MVesuviusC-curl_handle_setform_fix

Update search.R

MVesuviusC commented:
So this fixed the initial issue but later in the process another error pops up. I think it's the same sort of issue, the problem seems to be with sequence retrieval step. I assume that the function is being passed a variable of the wrong type (factor instead of numeric). I don't have time today to keep working on this but I'll try to circle back when I can.


`
> test <- search_primer_pair('CGGTTGGGGTGACCTCGGA', 'GCTGTTATCCCTAGGGTAACT')
BLASTing 1 primer combinations
No encoding supplied: defaulting to UTF-8.
Submitting Primer-BLAST query
No encoding supplied: defaulting to UTF-8.
BLAST alignment processing, refreshing in 20 seconds...
No encoding supplied: defaulting to UTF-8.
^[[6~BLAST alignment completed in 47 seconds
No encoding supplied: defaulting to UTF-8.
taxonomy retrieved in 24 seconds
Error in curl::handle_setform(handle, .list = req$fields) :
  Unsupported value type for form field 'id'.
Error in curl::handle_setform(handle, .list = req$fields) :
  Unsupported value type for form field 'id'.
Error in curl::handle_setform(handle, .list = req$fields) :
  Unsupported value type for form field 'id'.
`


`
>traceback()
32: body_config(body, match.arg(encode))
31: request_build("POST", hu$url, body_config(body, match.arg(encode)),
        as.request(config), ...)
30: fun(...)
29: POST_retry(fetch_url, body = query)
28: get_sequence(gi[i], start[i], stop[i])
27: f(...)
26: doTryCatch(return(expr), name, parentenv, handler)
25: tryCatchOne(expr, names, parentenv, handlers[[1L]])
24: tryCatchList(expr, classes, parentenv, handlers)
23: tryCatch(expr, error = function(e) {
        call <- conditionCall(e)
        if (!is.null(call)) {
            if (identical(call[[1L]], quote(doTryCatch)))
                call <- sys.call(-4L)
            dcall <- deparse(call)[1L]
            prefix <- paste("Error in", dcall, ": ")
            LONG <- 75L
            msg <- conditionMessage(e)
            sm <- strsplit(msg, "\n")[[1L]]
            w <- 14L + nchar(dcall, type = "w") + nchar(sm[1L], type = "w")
            if (is.na(w))
                w <- 14L + nchar(dcall, type = "b") + nchar(sm[1L],
                    type = "b")
            if (w > LONG)
                prefix <- paste0(prefix, "\n  ")
        }
        else prefix <- "Error : "
        msg <- paste0(prefix, conditionMessage(e), "\n")
        .Internal(seterrmessage(msg[1L]))
        if (!silent && identical(getOption("show.error.messages"),
            TRUE)) {
            cat(msg, file = stderr())
            .Internal(printDeferredWarnings())
        }
        invisible(structure(msg, class = "try-error", condition = e))
    })
22: try(result <- expr)
21: try_default(f(...), default, quiet = quiet)
20: .fun(piece, ...)
19: (function (i)
    {
        piece <- pieces[[i]]
        if (.inform) {
            res <- try(.fun(piece, ...))
            if (inherits(res, "try-error")) {
                piece <- paste(utils::capture.output(print(piece)),
                    collapse = "\n")
                stop("with piece ", i, ": \n", piece, call. = FALSE)
            }
        }
        else {
            res <- .fun(piece, ...)
        }
        progress$step()
        res
    })(314L)
18: .Call(loop_apply_, as.integer(n), f, env)
17: loop_apply(n, do.ply)
16: llply(.data = pieces, .fun = .fun, ..., .progress = .progress,
        .inform = .inform, .parallel = .parallel, .paropts = .paropts)
15: alply(seq_along(gi), .margins = 1, .parallel = .parallel, .progress = .progress,
        failwith(NA, f = get_sequence_itr))
14: get_sequences(primer_search$BLAST_result$gi, primer_search$BLAST_result$product_start,
        primer_search$BLAST_result$product_stop, simplify = simplify,
        .parallel = .parallel)
13: doTryCatch(return(expr), name, parentenv, handler)
12: tryCatchOne(expr, names, parentenv, handlers[[1L]])
11: tryCatchList(expr, classes, parentenv, handlers)
10: tryCatch(expr, error = function(e) {
        call <- conditionCall(e)
        if (!is.null(call)) {
            if (identical(call[[1L]], quote(doTryCatch)))
                call <- sys.call(-4L)
            dcall <- deparse(call)[1L]
            prefix <- paste("Error in", dcall, ": ")
            LONG <- 75L
            msg <- conditionMessage(e)
            sm <- strsplit(msg, "\n")[[1L]]
            w <- 14L + nchar(dcall, type = "w") + nchar(sm[1L], type = "w")
            if (is.na(w))
                w <- 14L + nchar(dcall, type = "b") + nchar(sm[1L],
                    type = "b")
            if (w > LONG)
                prefix <- paste0(prefix, "\n  ")
        }
        else prefix <- "Error : "
        msg <- paste0(prefix, conditionMessage(e), "\n")
        .Internal(seterrmessage(msg[1L]))
        if (!silent && identical(getOption("show.error.messages"),
            TRUE)) {
            cat(msg, file = stderr())
            .Internal(printDeferredWarnings())
        }
        invisible(structure(msg, class = "try-error", condition = e))
    })
9: try(result <- expr)
8: try_default({
       primer_search$name = if (!is.null(name))
           name
       else name = paste(forward, reverse, sep = "_")
       primer_search$arguments = c(forward = forward, reverse = reverse,
           name = name, num_aligns = num_aligns, num_permutations = num_permutations,
           simplify = simplify, clustal_options = clustal_options,
           list(...))
       primer_search$response = primer_search(forward, reverse,
           num_permutations = num_permutations, .progress = .progress,
           .parallel = .parallel, num_aligns = num_aligns, ...)
       start_time = now()
       primer_search$BLAST_result = filter_duplicates(ldply(primer_search$response,
           parse_primer_hits, .parallel = .parallel))
       message(nrow(primer_search$BLAST_result), " BLAST alignments parsed in ",
           seconds_elapsed_text(start_time))
       start_time = now()
       primer_search$taxonomy = get_taxonomy(primer_search$BLAST_result$gi)
       message("taxonomy retrieved in ", seconds_elapsed_text(start_time))
       start_time = now()
       primer_search$sequence = get_sequences(primer_search$BLAST_result$gi,
           primer_search$BLAST_result$product_start, primer_search$BLAST_result$product_stop,
           simplify = simplify, .parallel = .parallel)
       lengths = laply(primer_search$sequence, length)
       message(length(primer_search$sequence), " sequences retrieved from NCBI",
           " in ", seconds_elapsed_text(start_time), ", product length",
           " min:", min(lengths), " mean:", round(mean(lengths),
               2), " max:", max(lengths))
       start_time = now()
       primer_search$alignment = do.call(clustalo, c(list(primer_search$sequence,
           threads = getDoParWorkers()), clustal_options))
       message(nrow(primer_search$alignment), " sequences aligned in ",
           seconds_elapsed_text(start_time), " length:", ncol(primer_search$alignment))
       start_time = now()
       primer_search$distances = do.call(dist.dna, c(list(primer_search$alignment,
           distance_options)))
       message("pairwise DNA distances calculated in ", seconds_elapsed_text(start_time))
       start_time = now()
       primer_search$tree = tree_from_alignment(primer_search$alignment)
       message("constructed neighbor joining tree in ", seconds_elapsed_text(start_time))
       primer_search
   }, default = primer_search)
7: doTryCatch(return(expr), name, parentenv, handler)
6: tryCatchOne(expr, names, parentenv, handlers[[1L]])
5: tryCatchList(expr, classes, parentenv, handlers)
4: tryCatch(name, error = function(e) e)
3: ls(env)
2: env2list(try_default({
       primer_search$name = if (!is.null(name))
           name
       else name = paste(forward, reverse, sep = "_")
       primer_search$arguments = c(forward = forward, reverse = reverse,
           name = name, num_aligns = num_aligns, num_permutations = num_permutations,
           simplify = simplify, clustal_options = clustal_options,
           list(...))
       primer_search$response = primer_search(forward, reverse,
           num_permutations = num_permutations, .progress = .progress,
           .parallel = .parallel, num_aligns = num_aligns, ...)
       start_time = now()
       primer_search$BLAST_result = filter_duplicates(ldply(primer_search$response,
           parse_primer_hits, .parallel = .parallel))
       message(nrow(primer_search$BLAST_result), " BLAST alignments parsed in ",
           seconds_elapsed_text(start_time))
       start_time = now()
       primer_search$taxonomy = get_taxonomy(primer_search$BLAST_result$gi)
       message("taxonomy retrieved in ", seconds_elapsed_text(start_time))
       start_time = now()
       primer_search$sequence = get_sequences(primer_search$BLAST_result$gi,
           primer_search$BLAST_result$product_start, primer_search$BLAST_result$product_stop,
           simplify = simplify, .parallel = .parallel)
       lengths = laply(primer_search$sequence, length)
       message(length(primer_search$sequence), " sequences retrieved from NCBI",
           " in ", seconds_elapsed_text(start_time), ", product length",
           " min:", min(lengths), " mean:", round(mean(lengths),
               2), " max:", max(lengths))
       start_time = now()
       primer_search$alignment = do.call(clustalo, c(list(primer_search$sequence,
           threads = getDoParWorkers()), clustal_options))
       message(nrow(primer_search$alignment), " sequences aligned in ",
           seconds_elapsed_text(start_time), " length:", ncol(primer_search$alignment))
       start_time = now()
       primer_search$distances = do.call(dist.dna, c(list(primer_search$alignment,
           distance_options)))
       message("pairwise DNA distances calculated in ", seconds_elapsed_text(start_time))
       start_time = now()
       primer_search$tree = tree_from_alignment(primer_search$alignment)
       message("constructed neighbor joining tree in ", seconds_elapsed_text(start_time))
       primer_search
   }, default = primer_search))
1: search_primer_pair("CGGTTGGGGTGACCTCGGA", "GCTGTTATCCCTAGGGTAACT")
`

`
> sessionInfo()
R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS
locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] primerTree_1.0.3        devtools_1.13.4         gridExtra_2.3
[4] directlabels_2017.03.31

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.14     magrittr_1.5     munsell_0.4.3    colorspace_1.3-2
 [5] ape_5.0          lattice_0.20-35  R6_2.2.2         rlang_0.1.6
 [9] quadprog_1.5-5   foreach_1.4.4    stringr_1.2.0    httr_1.3.1
[13] plyr_1.8.4       tools_3.2.3      parallel_3.2.3   grid_3.2.3
[17] gtable_0.2.0     nlme_3.1-131     git2r_0.21.0     withr_2.1.1
[21] iterators_1.0.9  digest_0.6.14    lazyeval_0.2.1   tibble_1.4.1
[25] reshape2_1.4.3   ggplot2_2.2.1    codetools_0.2-15 curl_3.1
[29] memoise_1.1.0    stringi_1.1.6    pillar_1.1.0     scales_0.5.0
[33] XML_3.98-1.9     lubridate_1.7.1
`

jimhester commented on R/search.R:
```diff
@@ -48,7 +48,8 @@ enumerate_primers = function(forward, reverse){
              reverse=rep(enumerate_ambiguity(reverse),
-                         each=length(forward_primers)))
+                         each=length(forward_primers)),
+             stringsAsFactors = F)
```
This should really be `stringsAsFactors = FALSE`. `F` is just a variable that happens to be set to `FALSE`, it can actually be assigned any value. `FALSE` is a keyword which cannot be assigned.

``` r
F <- TRUE
if (F) print("hi")
#> [1] "hi"
FALSE <- TRUE
#> Error in FALSE <- TRUE: invalid (do_set) left-hand side to assignment
```

Created on 2018-01-17 by the [reprex package](http://reprex.tidyverse.org) (v0.1.1.9000).

MVesuviusC replied:
Good to know, I'll fix that.

MVesuviusC committed:
changed F to FALSE

Changed F to FALSE

MVesuviusC commented:
Not yet, I need to go back through to find what is causing the second error. During the sequence retrieval step I think there is a similar error to the first one I fixed where the variable being passed to curl to interface with the website is of the wrong variable type. I believe that newer versions of curl may be enforcing proper variable types instead of coercing them into what the website needs. The problem I fixed with my first edit here was a variable being a factor instead of a character. I haven't had time to play around to find exactly where the problem is yet.

MVesuviusC committed:
Fix for second curl error

I hope

MVesuviusC committed:
Merge pull request #2 from MVesuviusC/MVesuviusC-curl_handle_setform_fix

Fix for second curl error
